File: src/main/java/minseok/cqrschallenge/product/mapper/ProductMapper.java
package minseok.cqrschallenge.product.mapper;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import minseok.cqrschallenge.brand.entity.Brand;
import minseok.cqrschallenge.brand.repository.BrandRepository;
import minseok.cqrschallenge.category.repository.CategoryRepository;
import minseok.cqrschallenge.common.exception.ResourceNotFoundException;
import minseok.cqrschallenge.product.dto.request.ProductCreateRequest;
import minseok.cqrschallenge.product.dto.response.ProductCreateResponse;
import minseok.cqrschallenge.product.dto.response.ProductDetailResponse;
import minseok.cqrschallenge.product.dto.response.ProductListResponse;
import minseok.cqrschallenge.product.dto.response.ProductUpdateResponse;
import minseok.cqrschallenge.product.entity.Dimensions;
import minseok.cqrschallenge.product.entity.Product;
import minseok.cqrschallenge.product.entity.ProductCategory;
import minseok.cqrschallenge.product.entity.ProductDetail;
import minseok.cqrschallenge.product.entity.ProductImage;
import minseok.cqrschallenge.product.entity.ProductOption;
import minseok.cqrschallenge.product.entity.ProductOptionGroup;
import minseok.cqrschallenge.product.entity.ProductPrice;
import minseok.cqrschallenge.product.entity.ProductStatus;
import minseok.cqrschallenge.product.entity.ProductTag;
import minseok.cqrschallenge.product.repository.ProductOptionRepository;
import minseok.cqrschallenge.review.repository.ReviewRepository;
import minseok.cqrschallenge.seller.entity.Seller;
import minseok.cqrschallenge.seller.repository.SellerRepository;
import minseok.cqrschallenge.tag.entity.Tag;
import minseok.cqrschallenge.tag.repository.TagRepository;
import org.springframework.stereotype.Component;


@Component
@RequiredArgsConstructor
public class ProductMapper {

    private final ReviewRepository reviewRepository;
    private final BrandRepository brandRepository;
    private final SellerRepository sellerRepository;
    private final CategoryRepository categoryRepository;
    private final TagRepository tagRepository;
    private final ProductOptionRepository productOptionRepository;

    public Product toEntity(ProductCreateRequest request) {
        Product product = Product.builder()
            .name(request.getName())
            .slug(request.getSlug())
            .shortDescription(request.getShortDescription())
            .fullDescription(request.getFullDescription())
            .status(ProductStatus.valueOf(request.getStatus()))
            .build();

        mapSeller(product, request);
        mapBrand(product, request);
        mapPrice(product, request);
        mapDetail(product, request);
        mapCategories(product, request);
        mapOptionGroups(product, request);
        mapTags(product, request);

        return product;
    }

    public void mapImages(Product product, ProductCreateRequest request) {
        if (request.getImages() != null && !request.getImages().isEmpty()) {
            request.getImages().forEach(imageRequest -> {
                ProductImage.ProductImageBuilder imageBuilder = ProductImage.builder()
                    .url(imageRequest.getUrl())
                    .altText(imageRequest.getAltText())
                    .displayOrder(imageRequest.getDisplayOrder())
                    .isPrimary(imageRequest.getIsPrimary());

                if (imageRequest.getOptionId() != null) {
                    ProductOption option = productOptionRepository.findById(
                            imageRequest.getOptionId())
                        .orElseThrow(() -> new ResourceNotFoundException(
                            "옵션을 찾을 수 없습니다: " + imageRequest.getOptionId()));
                    imageBuilder.option(option);
                }

                ProductImage image = imageBuilder.build();
                product.addImage(image);
            });
        }
    }

    public ProductListResponse toListResponse(Product product) {
        return ProductListResponse.builder()
            .id(product.getId())
            .name(product.getName())
            .slug(product.getSlug())
            .seller(buildSellerSummary(product))
            .brand(buildBrandSummary(product))
            .shortDescription(product.getShortDescription())
            .basePrice(product.getPrice().getBasePrice())
            .salePrice(product.getPrice().getSalePrice())
            .currency(product.getPrice().getCurrency())
            .status(product.getStatus().name())
            .rating(getRating(product))
            .reviewCount(getReviewCount(product))
            .createdAt(product.getCreatedAt())
            .build();
    }

    public ProductDetailResponse toDetailResponse(Product product) {
        return ProductDetailResponse.builder()
            .id(product.getId())
            .name(product.getName())
            .slug(product.getSlug())
            .shortDescription(product.getShortDescription())
            .fullDescription(product.getFullDescription())
            .seller(buildSellerDetail(product))
            .brand(buildBrandDetail(product))
            .price(buildPriceInfo(product))
            .categories(buildCategoryInfo(product))
            .optionGroups(buildOptionGroups(product))
            .images(buildImages(product))
            .tags(buildTags(product))
            .detail(buildProductDetailInfo(product))
            .rating(buildRatingInfo(product))
            .status(product.getStatus().name())
            .createdAt(product.getCreatedAt())
            .updatedAt(product.getUpdatedAt())
            .build();
    }


    public ProductCreateResponse toCreateResponse(Product product) {
        return ProductCreateResponse.builder()
            .id(product.getId())
            .name(product.getName())
            .slug(product.getSlug())
            .createdAt(product.getCreatedAt())
            .updatedAt(product.getUpdatedAt())
            .build();
    }


    public ProductUpdateResponse toUpdateResponse(Product product) {
        return ProductUpdateResponse.builder()
            .id(product.getId())
            .name(product.getName())
            .slug(product.getSlug())
            .updatedAt(product.getUpdatedAt())
            .build();
    }

    private void mapSeller(Product product, ProductCreateRequest request) {
        if (request.getSellerId() != null) {
            Seller seller = sellerRepository.findById(request.getSellerId())
                .orElseThrow(() -> new ResourceNotFoundException("판매자를 찾을 수 없습니다."));
            product.assignSeller(seller);
        }
    }

    private void mapBrand(Product product, ProductCreateRequest request) {
        if (request.getBrandId() != null) {
            Brand brand = brandRepository.findById(request.getBrandId())
                .orElseThrow(() -> new ResourceNotFoundException("브랜드를 찾을 수 없습니다."));
            product.assignBrand(brand);
        }
    }

    private void mapPrice(Product product, ProductCreateRequest request) {
        if (request.getPrice() != null) {
            product.addPrice(
                ProductPrice.builder()
                    .basePrice(request.getPrice().getBasePrice())
                    .salePrice(request.getPrice().getSalePrice())
                    .costPrice(request.getPrice().getCostPrice())
                    .currency(request.getPrice().getCurrency())
                    .taxRate(request.getPrice().getTaxRate())
                    .build()
            );
        }
    }

    private void mapDetail(Product product, ProductCreateRequest request) {
        if (request.getDetail() != null) {
            product.addDetail(ProductDetail.builder()
                .product(product)
                .weight(request.getDetail().getWeight())
                .dimensions(request.getDetail().getDimensions())
                .materials(request.getDetail().getMaterials())
                .countryOfOrigin(request.getDetail().getCountryOfOrigin())
                .warrantyInfo(request.getDetail().getWarrantyInfo())
                .careInstructions(request.getDetail().getCareInstructions())
                .additionalInfo(request.getDetail().getAdditionalInfo())
                .build());
        }
    }

    private void mapCategories(Product product, ProductCreateRequest request) {
        if (request.getCategories() != null && !request.getCategories().isEmpty()) {
            request.getCategories().forEach(categoryRequest ->
                product.addCategory(ProductCategory.builder()
                    .category(categoryRepository.findById(categoryRequest.getCategoryId())
                        .orElseThrow(() -> new ResourceNotFoundException("카테고리를 찾을 수 없습니다.")))
                    .isPrimary(categoryRequest.getIsPrimary())
                    .build()));
        }
    }

    private void mapOptionGroups(Product product, ProductCreateRequest request) {
        if (request.getOptionGroups() != null && !request.getOptionGroups().isEmpty()) {
            request.getOptionGroups().forEach(optionGroupRequest -> {
                ProductOptionGroup optionGroup = ProductOptionGroup.builder()
                    .name(optionGroupRequest.getName())
                    .displayOrder(optionGroupRequest.getDisplayOrder())
                    .build();
                product.addOptionGroup(optionGroup);

                optionGroupRequest.getOptions().forEach(optionRequest -> {
                    ProductOption option = ProductOption.builder()
                        .name(optionRequest.getName())
                        .additionalPrice(optionRequest.getAdditionalPrice())
                        .sku(optionRequest.getSku())
                        .stock(optionRequest.getStock())
                        .displayOrder(optionRequest.getDisplayOrder())
                        .build();
                    optionGroup.addOption(option);
                });
            });
        }
    }

    private void mapTags(Product product, ProductCreateRequest request) {
        if (request.getTags() != null && !request.getTags().isEmpty()) {
            request.getTags().forEach(tagId -> {
                Tag tag = tagRepository.findById(tagId)
                    .orElseThrow(() -> new ResourceNotFoundException("태그를 찾을 수 없습니다."));
                ProductTag productTag = ProductTag.builder()
                    .tag(tag)
                    .build();
                product.addTag(productTag);
            });
        }
    }

    private double getRating(Product product) {
        Long productId = product.getId();
        Integer reviewCount = reviewRepository.countByProductId(productId);

        if (reviewCount == 0) {
            return 0.0;
        } else {
            return reviewRepository.calculateAverageRating(productId);
        }
    }

    private int getReviewCount(Product product) {
        return reviewRepository.countByProductId(product.getId());
    }

    private ProductListResponse.BrandSummary buildBrandSummary(Product product) {
        if (product.getBrand() == null) {
            return null;
        }

        return ProductListResponse.BrandSummary.builder()
            .id(product.getBrand().getId())
            .name(product.getBrand().getName())
            .build();
    }

    private ProductListResponse.SellerSummary buildSellerSummary(Product product) {
        if (product.getSeller() == null) {
            return null;
        }

        return ProductListResponse.SellerSummary.builder()
            .id(product.getSeller().getId())
            .name(product.getSeller().getName())
            .build();
    }


    private ProductDetailResponse.SellerDetail buildSellerDetail(Product product) {
        if (product.getSeller() == null) {
            return null;
        }

        return ProductDetailResponse.SellerDetail.builder()
            .id(product.getSeller().getId())
            .name(product.getSeller().getName())
            .description(product.getSeller().getDescription())
            .logoUrl(product.getSeller().getLogoUrl())
            .rating(product.getSeller().getRating())
            .contactEmail(product.getSeller().getContactEmail())
            .contactPhone(product.getSeller().getContactPhone())
            .build();
    }

    private ProductDetailResponse.ProductDetailInfo buildProductDetailInfo(Product product) {
        if (product.getDetail() == null) {
            return null;
        }

        return ProductDetailResponse.ProductDetailInfo.builder()
            .dimensions(
                new Dimensions(product.getDetail().getDimensions().getHeight(),
                    product.getDetail().getDimensions().getWidth(),
                    product.getDetail().getDimensions().getHeight()))
            .weight(product.getDetail().getWeight())
            .countryOfOrigin(product.getDetail().getCountryOfOrigin())
            .materials(product.getDetail().getMaterials())
            .warrantyInfo(product.getDetail().getWarrantyInfo())
            .careInstructions(product.getDetail().getCareInstructions())
            .additionalInfo(product.getDetail().getAdditionalInfo())
            .build();
    }

    private ProductDetailResponse.BrandDetail buildBrandDetail(Product product) {
        if (product.getBrand() == null) {
            return null;
        }

        return ProductDetailResponse.BrandDetail.builder()
            .id(product.getBrand().getId())
            .name(product.getBrand().getName())
            .description(product.getBrand().getDescription())
            .logoUrl(product.getBrand().getLogoUrl())
            .website(product.getBrand().getWebsite())
            .build();
    }


    private ProductDetailResponse.ProductPriceInfo buildPriceInfo(Product product) {
        if (product.getPrice() == null) {
            return null;
        }

        BigDecimal basePrice = product.getPrice().getBasePrice();
        BigDecimal salePrice = product.getPrice().getSalePrice();

        Integer discountPercentage = null;
        if (basePrice != null && salePrice != null && basePrice.compareTo(BigDecimal.ZERO) > 0) {
            discountPercentage = basePrice
                .subtract(salePrice)
                .multiply(new BigDecimal("100"))
                .divide(basePrice, 0, RoundingMode.HALF_UP)
                .intValue();
        }

        return ProductDetailResponse.ProductPriceInfo.builder()
            .basePrice(basePrice)
            .salePrice(salePrice)
            .currency(product.getPrice().getCurrency())
            .taxRate(product.getPrice().getTaxRate())
            .discountPercentage(discountPercentage)
            .build();
    }

    private List<ProductDetailResponse.CategoryInfo> buildCategoryInfo(Product product) {
        if (product.getCategories() == null || product.getCategories().isEmpty()) {
            return null;
        }

        return product.getCategories().stream()
            .map(category -> ProductDetailResponse.CategoryInfo.builder()
                .id(category.getId())
                .name(category.getCategory().getName())
                .slug(category.getCategory().getSlug())
                .isPrimary(category.getIsPrimary())
                .parent(category.getCategory().getParent() != null
                    ? ProductDetailResponse.CategoryInfo.CategoryParentInfo.builder()
                    .id(category.getCategory().getParent().getId())
                    .name(category.getCategory().getParent().getName())
                    .slug(category.getCategory().getParent().getSlug())
                    .build()
                    : null)
                .build())
            .collect(Collectors.toList());
    }

    private List<ProductDetailResponse.OptionGroupInfo> buildOptionGroups(Product product) {
        if (product.getOptionGroups() == null || product.getOptionGroups().isEmpty()) {
            return null;
        }

        return product.getOptionGroups().stream()
            .map(og -> ProductDetailResponse.OptionGroupInfo.builder()
                .id(og.getId())
                .name(og.getName())
                .displayOrder(og.getDisplayOrder())
                .options(og.getOptions().stream()
                    .map(opt -> ProductDetailResponse.OptionGroupInfo.OptionInfo.builder()
                        .id(opt.getId())
                        .name(opt.getName())
                        .additionalPrice(opt.getAdditionalPrice())
                        .sku(opt.getSku())
                        .stock(opt.getStock())
                        .displayOrder(opt.getDisplayOrder())
                        .build())
                    .collect(Collectors.toList()))
                .build())
            .collect(Collectors.toList());
    }

    private List<ProductDetailResponse.ProductImageInfo> buildImages(Product product) {
        if (product.getImages() == null || product.getImages().isEmpty()) {
            return null;
        }

        return product.getImages().stream()
            .map(img -> new ProductDetailResponse.ProductImageInfo(
                img.getId(),
                img.getUrl(),
                img.getAltText(),
                img.getIsPrimary(),
                img.getDisplayOrder(),
                img.getOption() != null ? img.getOption().getId() : null))
            .collect(Collectors.toList());
    }

    private List<ProductDetailResponse.TagInfo> buildTags(Product product) {
        if (product.getTags() == null || product.getTags().isEmpty()) {
            return null;
        }

        return product.getTags().stream()
            .map(tag -> new ProductDetailResponse.TagInfo(
                tag.getId(),
                tag.getTag().getName(),
                tag.getTag().getSlug()))
            .collect(Collectors.toList());
    }

    private ProductDetailResponse.RatingInfo buildRatingInfo(Product product) {
        Long productId = product.getId();
        Integer reviewCount = reviewRepository.countByProductId(productId);
        if (reviewCount == 0) {
            return null;
        }

        Double averageRating = reviewRepository.calculateAverageRating(productId);

        List<Object[]> ratingDistribution = reviewRepository.countByRatingForProduct(productId);
        Map<String, Integer> distribution = new HashMap<>();

        for (int i = 1; i <= 5; i++) {
            distribution.put(String.valueOf(i), 0);
        }

        for (Object[] data : ratingDistribution) {
            Integer rating = ((Number) data[0]).intValue();
            Integer count = ((Number) data[1]).intValue();
            distribution.put(String.valueOf(rating), count);
        }

        return ProductDetailResponse.RatingInfo.builder()
            .average(averageRating)
            .count(reviewCount)
            .distribution(distribution)
            .build();
    }
}

File: src/types/mappers/product.mapper.ts
import {
  Prisma,
  Product,
  ProductDetail,
  ProductImage,
  ProductOption,
  ProductOptionGroup,
} from '@prisma/client';
import {
  ProductDetailResponse,
  ProductImageResponse,
  ProductListItemResponse,
  ProductOptionModifyResponse,
  ProductOptionResponse,
} from '../responses/product.response';

// 추가: JSON 차원 정보에 대한 인터페이스 정의
interface ProductDimensions {
  width: number;
  height: number;
  depth: number;
}

// 확장된 Product 타입 정의 (모든 include된 관계를 포함)
export type ProductWithRelations = Product & {
  seller?: {
    id: number;
    name: string;
    description?: string;
    logoUrl?: string;
    rating?: number;
    contactEmail?: string;
    contactPhone?: string;
  };
  brand: {
    id: number;
    name: string;
    slug: string;
    description?: string;
    logoUrl?: string;
    website?: string;
  };
  detail?: ProductDetail;
  price?: {
    basePrice: number;
    salePrice?: number;
    currency: string;
    taxRate?: number;
    costPrice?: number;
  };
  images: ProductImage[];
  optionGroups?: (ProductOptionGroup & {
    options?: ProductOption[];
  })[];
  categories?: {
    isPrimary: boolean;
    category: {
      id: number;
      name: string;
      slug: string;
      parentId?: number;
      parent?: {
        id: number;
        name: string;
        slug: string;
      };
    };
  }[];
  tags?: {
    tag: {
      id: number;
      name: string;
      slug: string;
    };
  }[];
  rating?: {
    average: number;
    count: number;
    distribution: Record<string, number>;
  };
  relatedProducts?: ProductWithRelations[];
};

// 추가: additionalInfo의 타입 가드 함수
function isValidAdditionalInfo(
  obj: unknown,
): obj is Record<string, string | number | boolean | null> {
  return obj !== null && typeof obj === 'object';
}

/**
 * 상품 목록 항목으로 변환하는 매퍼
 */
export function mapProductToListItem(product: ProductWithRelations): ProductListItemResponse {
  const primaryImage = product.images.find((img) => img.isPrimary);

  return {
    id: product.id,
    name: product.name,
    slug: product.slug,
    short_description: product.shortDescription || undefined,
    base_price: product.price?.basePrice || 0,
    sale_price: product.price?.salePrice || undefined,
    currency: product.price?.currency || 'KRW',
    primary_image: primaryImage
      ? {
          url: primaryImage.url,
          alt_text: primaryImage.altText || undefined,
        }
      : undefined,
    brand: {
      id: product.brand.id,
      name: product.brand.name,
    },
    seller: {
      id: product.seller?.id || product.sellerId,
      name: product.seller?.name || '판매자',
    },
    rating: product.rating?.average,
    review_count: product.rating?.count,
    in_stock:
      product.optionGroups?.some((group) => group.options?.some((option) => option.stock > 0)) ??
      false,
    status: product.status,
    created_at: product.createdAt.toISOString(),
  };
}

/**
 * 상품 상세 정보로 변환하는 매퍼
 */
export function mapProductToDetail(product: ProductWithRelations): ProductDetailResponse {
  return {
    id: product.id,
    name: product.name,
    slug: product.slug,
    short_description: product.shortDescription || undefined,
    full_description: product.fullDescription || undefined,
    seller: {
      id: product.seller?.id || product.sellerId,
      name: product.seller?.name || '판매자',
      description: product.seller?.description,
      logo_url: product.seller?.logoUrl,
      rating: product.seller?.rating,
      contact_email: product.seller?.contactEmail,
      contact_phone: product.seller?.contactPhone,
    },
    brand: {
      id: product.brand.id,
      name: product.brand.name,
      slug: product.brand.slug,
      description: product.brand.description,
      logo_url: product.brand.logoUrl,
      website: product.brand.website,
    },
    status: product.status,
    created_at: product.createdAt.toISOString(),
    updated_at: product.updatedAt.toISOString(),
    detail: {
      weight: product.detail?.weight ?? undefined,
      dimensions: product.detail?.dimensions
        ? (() => {
            const dim = product.detail.dimensions;
            // Prisma에서 JSON 형태로 저장된 dimensions를 적절히 변환
            if (
              dim &&
              typeof dim === 'object' &&
              'width' in dim &&
              'height' in dim &&
              'depth' in dim
            ) {
              return {
                width: Number(dim.width),
                height: Number(dim.height),
                depth: Number(dim.depth),
              };
            }
            return undefined;
          })()
        : undefined,
      materials: product.detail?.materials ?? undefined,
      country_of_origin: product.detail?.countryOfOrigin ?? undefined,
      warranty_info: product.detail?.warrantyInfo ?? undefined,
      care_instructions: product.detail?.careInstructions ?? undefined,
      additional_info:
        product.detail?.additionalInfo && isValidAdditionalInfo(product.detail.additionalInfo)
          ? product.detail.additionalInfo
          : undefined,
    },
    price: {
      base_price: product.price?.basePrice || 0,
      sale_price: product.price?.salePrice,
      currency: product.price?.currency || 'KRW',
      tax_rate: product.price?.taxRate,
      discount_percentage:
        product.price?.salePrice && product.price.basePrice
          ? Math.round((1 - product.price.salePrice / product.price.basePrice) * 100)
          : undefined,
    },
    categories:
      product.categories?.map((cat) => ({
        id: cat.category.id,
        name: cat.category.name,
        slug: cat.category.slug,
        is_primary: cat.isPrimary || false,
        parent: cat.category.parentId
          ? {
              id: cat.category.parent?.id || 0,
              name: cat.category.parent?.name || '',
              slug: cat.category.parent?.slug || '',
            }
          : undefined,
      })) || [],
    option_groups:
      product.optionGroups?.map((group) => ({
        id: group.id,
        name: group.name,
        display_order: group.displayOrder,
        options: group.options?.map((opt) => mapOptionToResponse(opt)) || [],
      })) || [],
    images: product.images.map((img) => mapImageToResponse(img)),
    tags:
      product.tags?.map((tagRel) => ({
        id: tagRel.tag.id,
        name: tagRel.tag.name,
        slug: tagRel.tag.slug,
      })) || [],
    rating: {
      average: product.rating?.average || 0,
      count: product.rating?.count || 0,
      distribution: product.rating?.distribution || {
        '5': 0,
        '4': 0,
        '3': 0,
        '2': 0,
        '1': 0,
      },
    },
    related_products:
      product.relatedProducts?.map((related) => ({
        id: related.id,
        name: related.name,
        slug: related.slug,
        short_description: related.shortDescription ?? undefined,
        primary_image: related.images.find((img) => img.isPrimary)
          ? {
              url: related.images.find((img) => img.isPrimary)?.url || '',
              alt_text: related.images.find((img) => img.isPrimary)?.altText ?? undefined,
            }
          : undefined,
        base_price: related.price?.basePrice || 0,
        sale_price: related.price?.salePrice,
        currency: related.price?.currency || 'KRW',
      })) || [],
  };
}

/**
 * 상품 옵션 정보로 변환하는 매퍼
 */
export function mapOptionToResponse(option: ProductOption): ProductOptionResponse {
  return {
    id: option.id,
    name: option.name,
    additional_price: option.additionalPrice || 0,
    sku: option.sku || undefined,
    stock: option.stock,
    display_order: option.displayOrder,
  };
}

/**
 * 수정용 상품 옵션 정보로 변환하는 매퍼
 */
export function mapOptionToModifyResponse(option: ProductOption): ProductOptionModifyResponse {
  return {
    id: option.id,
    option_group_id: option.optionGroupId,
    name: option.name,
    additional_price: option.additionalPrice || 0,
    sku: option.sku || undefined,
    stock: option.stock,
    display_order: option.displayOrder,
  };
}

/**
 * 상품 이미지 정보로 변환하는 매퍼
 */
export function mapImageToResponse(image: ProductImage): ProductImageResponse {
  return {
    id: image.id,
    url: image.url,
    alt_text: image.altText || undefined,
    is_primary: image.isPrimary,
    display_order: image.displayOrder,
    option_id: image.optionId || undefined,
  };
}

/**
 * 리뷰 요약 정보 생성
 */
export function generateReviewSummary(reviews: { rating: number }[]) {
  const count = reviews.length;
  let average = 0;
  const distribution: Record<string, number> = {
    '5': 0,
    '4': 0,
    '3': 0,
    '2': 0,
    '1': 0,
  };

  if (count > 0) {
    reviews.forEach((review) => {
      distribution[review.rating.toString()] = (distribution[review.rating.toString()] || 0) + 1;
      average += review.rating;
    });

    average = Math.round((average / count) * 10) / 10;
  }

  return {
    average,
    count,
    distribution,
  };
}
